<launch>
  <!-- Python 인터프리터 강제 (ROS는 launch-prefix로 선행 실행) -->
  <arg name="python" default="/opt/conda/envs/python310/bin/python3"/>

  <!-- 토픽/프레임 -->
  <arg name="camera_image_topic"   default="/camera/image"/>
  <arg name="cloud_topic_a"        default="/color_scan_rel"/>
  <arg name="cloud_topic_b"        default="/color_scan_relative"/>
  <arg name="prompt_topic"         default="/instruction"/>
  <arg name="map_frame"            default="map"/>

  <!-- 파일 출력 -->
  <arg name="dump_dir"             default="/home/$USER/align_viz_runs"/>

  <!-- Gemini -->
  <arg name="gemini_model"         default="gemini-2.5-flash"/>
  <arg name="seg_prompt"           default=""/>

  <!-- 카메라 파라미터 -->
  <param name="align_viz_node/camera/width"    value="1920"/>
  <param name="align_viz_node/camera/height"   value="640"/>
  <param name="align_viz_node/camera/hfov_deg" value="360.0"/>
  <param name="align_viz_node/camera/vfov_deg" value="120.0"/>

  <!-- LiDAR->Camera 외부 파라미터 (로봇 실측치 적용) -->
  <arg name="l2c_x"     default="-0.12"/>
  <arg name="l2c_y"     default="-0.075"/>
  <arg name="l2c_z"     default="0.265"/>
  <arg name="l2c_roll"  default="-1.5707963"/>
  <arg name="l2c_pitch" default="0.0"/>
  <arg name="l2c_yaw"   default="-1.5707963"/>

  <!-- LiDAR 파라미터(기본 0) -->
  <arg name="lidar_x"     default="0.0"/>
  <arg name="lidar_y"     default="0.0"/>
  <arg name="lidar_z"     default="0.0"/>
  <arg name="lidar_roll"  default="0.0"/>
  <arg name="lidar_pitch" default="0.0"/>
  <arg name="lidar_yaw"   default="0.0"/>

  <!-- 실행/타임아웃 -->
  <arg name="cooldown_sec"    default="10.0"/>
  <arg name="seg_timeout_sec" default="20.0"/>

  <!-- 정지 감지 자동 트리거 -->
  <arg name="auto_on_stop"        default="true"/>
  <arg name="odom_topic"          default="/odom"/>
  <arg name="stop_lin_thresh"     default="0.02"/>
  <arg name="stop_ang_thresh"     default="0.02"/>
  <arg name="stop_dwell_sec"      default="1.0"/>
  <arg name="auto_cooldown_sec"   default="8.0"/>
  <arg name="tf_stop_enable"      default="true"/>
  <arg name="tf_stop_parent"      default="map"/>
  <arg name="tf_stop_child"       default="base_link"/>

  <node pkg="gemini_api" type="align_viz_node.py" name="align_viz_node" output="screen"
        launch-prefix="$(arg python)">
    <param name="dump_dir"             value="/home/aailab/cwh316/CMU-VLA-Challenge/dumps_align_viz"/>
    <param name="map_frame"            value="$(arg map_frame)"/>
    <param name="prompt_topic"         value="$(arg prompt_topic)"/>
    <param name="gemini_model"         value="$(arg gemini_model)"/>
    <param name="seg_prompt"           value="$(arg seg_prompt)"/>

    <param name="camera_image_topic"   value="$(arg camera_image_topic)"/>
    <param name="cloud_topic_a"        value="$(arg cloud_topic_a)"/>
    <param name="cloud_topic_b"        value="$(arg cloud_topic_b)"/>

    <!-- camera dict -->
    <rosparam param="camera">
      width: 1920
      height: 640
      hfov_deg: 360.0
      vfov_deg: 120.0
    </rosparam>

    <!-- l2c.* -->
    <param name="l2c/x"     value="$(arg l2c_x)"/>
    <param name="l2c/y"     value="$(arg l2c_y)"/>
    <param name="l2c/z"     value="$(arg l2c_z)"/>
    <param name="l2c/roll"  value="$(arg l2c_roll)"/>
    <param name="l2c/pitch" value="$(arg l2c_pitch)"/>
    <param name="l2c/yaw"   value="$(arg l2c_yaw)"/>

    <!-- lidar.* -->
    <param name="lidar/x"     value="$(arg lidar_x)"/>
    <param name="lidar/y"     value="$(arg lidar_y)"/>
    <param name="lidar/z"     value="$(arg lidar_z)"/>
    <param name="lidar/roll"  value="$(arg lidar_roll)"/>
    <param name="lidar/pitch" value="$(arg lidar_pitch)"/>
    <param name="lidar/yaw"   value="$(arg lidar_yaw)"/>

    <!-- run-time -->
    <param name="cooldown_sec"    value="$(arg cooldown_sec)"/>
    <param name="seg_timeout_sec" value="$(arg seg_timeout_sec)"/>

    <!-- auto trigger -->
    <param name="auto_on_stop"        value="$(arg auto_on_stop)"/>
    <param name="odom_topic"          value="$(arg odom_topic)"/>
    <param name="stop_lin_thresh"     value="$(arg stop_lin_thresh)"/>
    <param name="stop_ang_thresh"     value="$(arg stop_ang_thresh)"/>
    <param name="stop_dwell_sec"      value="$(arg stop_dwell_sec)"/>
    <param name="auto_cooldown_sec"   value="$(arg auto_cooldown_sec)"/>
    <param name="tf_stop_enable"      value="$(arg tf_stop_enable)"/>
    <param name="tf_stop_parent"      value="$(arg tf_stop_parent)"/>
    <param name="tf_stop_child"       value="$(arg tf_stop_child)"/>
  </node>
</launch>
